infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/StateTransferBatchProcessor.java

```diff
@@ -43,53 +65,99 @@
      */
     default TransferBatchResponse writeRecords(ReadBatch readBatch, LogUnitClient logUnitClient,
                                                int writeRetriesAllowed, Duration writeSleepDuration) {
-        List<Long> totalAddressesInRequest = readBatch.getAddresses();
-        Optional<String> destination = readBatch.getDestination();
+        ImmutableList<LogData> dataToTransfer = ImmutableList.copyOf(readBatch.getData());
+        ImmutableList<Long> addressesToTransfer = ImmutableList.copyOf(readBatch.getAddresses());
         Optional<Exception> lastWriteException = Optional.empty();
         Optional<TransferBatchResponse> transferBatchResponse = Optional.empty();
 
         for (int i = 0; i < writeRetriesAllowed; i++) {
             List<LogData> remainingDataToWrite = readBatch.getData();
             try {
-                boolean writeSucceeded = logUnitClient.writeRange(remainingDataToWrite).join();
-                if (!writeSucceeded) {
-                    throw new IllegalStateException("Failed to write to a log unit server.");
-                } else {
-                    lastWriteException = Optional.empty();
+                CFUtils.getUninterruptibly(logUnitClient.writeRange(remainingDataToWrite),
+                        OverwriteException.class,
+                        TimeoutException.class,
+                        NetworkException.class);
+                transferBatchResponse = Optional.of(TransferBatchResponse
+                        .builder()
+                        .transferBatchRequest(new TransferBatchRequest(addressesToTransfer,
+                                readBatch.getDestinationNode().map(ImmutableList::of), DATA))
+                        .status(SUCCEEDED)
+                        .build());
+                break;
+            } catch (OverwriteException | TimeoutException | NetworkException e) {
```

Comment:
For OverwriteException, is it possible for us to check its cause?

Reply:
This is rangeWrite to the wiped out part of the log; the only cause for OverwirteException that it throws is SAME_DATA.